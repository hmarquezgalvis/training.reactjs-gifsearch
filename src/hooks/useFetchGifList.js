import { useEffect, useState } from "react";
import { getGifList } from "../helpers/getGifs";

export const useFetchGifList = ( category , limit = 12 ) => {

  const initialState = { items: [], loading: true };
  const [state, setState] = useState(initialState);

  useEffect( () => {
    getGifList( category, limit ).then(images => setState({
        items: images,
        loading: false,
      })
    );
  }, [ category, limit ]);

  return state;
};