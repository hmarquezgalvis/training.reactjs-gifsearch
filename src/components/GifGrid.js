import React from 'react'
import PropTypes from 'prop-types';

import { GifGridItem } from './GifGridItem';
import { useFetchGifList } from '../hooks/useFetchGifList';

const GifGrid = ({ category }) => {

  const { items: imageList, loading } = useFetchGifList( category );

  return (
    <div className="card mb-3">
      <div className="card-body">
        <h5 className="card-title">
          { category }
          { 
            loading && <div className="spinner-grow spinner-grow-sm ms-1" role="status">
                        <span className="visually-hidden">Loading...</span>
                      </div>
          }
        </h5>

        <div className="d-flex align-content-stretch flex-wrap">
          {
            imageList.map( (item, idx) =>
              <GifGridItem 
                key={ item.id }
                { ...item } />
            )
          }
        </div>
      </div>
    </div>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
};

export default GifGrid;
