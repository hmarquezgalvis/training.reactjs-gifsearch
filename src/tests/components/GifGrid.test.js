import { shallow } from 'enzyme';

import GifGrid from '../../components/GifGrid'
import { useFetchGifList } from '../../hooks/useFetchGifList'

jest.mock('../../hooks/useFetchGifList.js');

describe( 'tests for <GifGrid />', () => {

  const category = 'One Punch';

  test( 'must be the right html generated', () => {
    useFetchGifList.mockReturnValue({
      items: [],
      loading: true,
    });

    const wrapper = shallow( <GifGrid category={ category } />)

    expect( wrapper ).toMatchSnapshot();
  });

  test( 'must be show items when load images using useFetchGifList()', () => {
    const items = [
      {
        id: 'dyjrpqaUVqCELGuQVr',
        title: 'Kimetsu No Yaiba Demon Slayer GIF by Crunchyroll',
        url: 'https://media3.giphy.com/media/dyjrpqaUVqCELGuQVr/giphy-downsized-medium.gif?cid=46e040a5twrrjqmg74ppo6cc0cu8ijv6466ipnjffiezj0ca&rid=giphy-downsized-medium.gif&ct=g',
      },
      {
        id: 'dyjrpqaUVqCELGuQVr',
        title: 'Kimetsu No Yaiba Demon Slayer GIF by Crunchyroll',
        url: 'https://media3.giphy.com/media/dyjrpqaUVqCELGuQVr/giphy-downsized-medium.gif?cid=46e040a5twrrjqmg74ppo6cc0cu8ijv6466ipnjffiezj0ca&rid=giphy-downsized-medium.gif&ct=g',
      },
    ];

    useFetchGifList.mockReturnValue({
      items,
      loading: false,
    });

    const wrapper = shallow( <GifGrid category={ category } />);

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find('p').exists() ).toBe( false );
    expect( wrapper.find('GifGridItem').length ).toBe( 2 );

  });

});