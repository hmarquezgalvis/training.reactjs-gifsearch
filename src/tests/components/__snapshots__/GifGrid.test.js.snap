// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <GifGrid /> must be show items when load images using useFetchGifList() 1`] = `
<div
  className="card mb-3"
>
  <div
    className="card-body"
  >
    <h5
      className="card-title"
    >
      One Punch
    </h5>
    <div
      className="d-flex align-content-stretch flex-wrap"
    >
      <GifGridItem
        id="dyjrpqaUVqCELGuQVr"
        key="dyjrpqaUVqCELGuQVr"
        title="Kimetsu No Yaiba Demon Slayer GIF by Crunchyroll"
        url="https://media3.giphy.com/media/dyjrpqaUVqCELGuQVr/giphy-downsized-medium.gif?cid=46e040a5twrrjqmg74ppo6cc0cu8ijv6466ipnjffiezj0ca&rid=giphy-downsized-medium.gif&ct=g"
      />
      <GifGridItem
        id="dyjrpqaUVqCELGuQVr"
        key="dyjrpqaUVqCELGuQVr"
        title="Kimetsu No Yaiba Demon Slayer GIF by Crunchyroll"
        url="https://media3.giphy.com/media/dyjrpqaUVqCELGuQVr/giphy-downsized-medium.gif?cid=46e040a5twrrjqmg74ppo6cc0cu8ijv6466ipnjffiezj0ca&rid=giphy-downsized-medium.gif&ct=g"
      />
    </div>
  </div>
</div>
`;

exports[`tests for <GifGrid /> must be the right html generated 1`] = `
<div
  className="card mb-3"
>
  <div
    className="card-body"
  >
    <h5
      className="card-title"
    >
      One Punch
      <div
        className="spinner-grow spinner-grow-sm ms-1"
        role="status"
      >
        <span
          className="visually-hidden"
        >
          Loading...
        </span>
      </div>
    </h5>
    <div
      className="d-flex align-content-stretch flex-wrap"
    />
  </div>
</div>
`;
